internal/extender/unschedulablepods.go

```diff
@@ -0,0 +1,155 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package extender
+
+import (
+	"context"
+	"time"
+
+	"github.com/palantir/k8s-spark-scheduler-lib/pkg/resources"
+	"github.com/palantir/witchcraft-go-logging/wlog/svclog/svc1log"
+	"github.com/palantir/witchcraft-go-logging/wlog/wapp"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	podutil "k8s.io/kubernetes/pkg/api/v1/pod"
+)
+
+const (
+	podExceedsClusterCapacity       v1.PodConditionType = "PodExceedsClusterCapacity"
+	unschedulablePollingInterval    time.Duration       = time.Minute
+	unschedulableInClusterThreshold time.Duration       = 10 * time.Minute
+)
+
+// UnschedulablePodMarker checks for spark scheduler managed pending driver pods
+// and checks if they can fit if the cluster was empty, else marks them with a
+// custom pod condition.
+type UnschedulablePodMarker struct {
+	nodeLister       corelisters.NodeLister
+	podLister        corelisters.PodLister
+	coreClient       corev1.CoreV1Interface
+	overheadComputer *OverheadComputer
+	binpacker        *Binpacker
+}
+
+// NewUnschedulablePodMarker creates a new UnschedulablePodMarker
+func NewUnschedulablePodMarker(
+	nodeLister corelisters.NodeLister,
+	podLister corelisters.PodLister,
+	coreClient corev1.CoreV1Interface,
+	overheadComputer *OverheadComputer,
+	binpacker *Binpacker) *UnschedulablePodMarker {
+	return &UnschedulablePodMarker{
+		nodeLister:       nodeLister,
+		podLister:        podLister,
+		coreClient:       coreClient,
+		overheadComputer: overheadComputer,
+		binpacker:        binpacker,
+	}
+}
+
+// Start starts periodic scanning for unschedulable applications
+func (u *UnschedulablePodMarker) Start(ctx context.Context) {
+	_ = wapp.RunWithFatalLogging(ctx, u.doStart)
+}
+
+func (u *UnschedulablePodMarker) doStart(ctx context.Context) error {
+	t := time.NewTicker(unschedulablePollingInterval)
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-t.C:
+			u.scanForUnschedulablePods(ctx)
+		}
+	}
+}
+
+func (u *UnschedulablePodMarker) scanForUnschedulablePods(ctx context.Context) {
+	pods, err := u.podLister.List(labels.Everything())
+	if err != nil {
+		svc1log.FromContext(ctx).Error("failed to list pods", svc1log.Stacktrace(err))
+		return
+	}
+	now := time.Now()
+	for _, pod := range pods {
+		if pod.Spec.SchedulerName == SparkSchedulerName &&
+			len(pod.Spec.NodeName) == 0 &&
+			pod.DeletionTimestamp == nil &&
+			pod.Labels[SparkRoleLabel] == Driver &&
+			pod.CreationTimestamp.Time.Add(unschedulableInClusterThreshold).Before(now) {
+
+			exceedsCapacity, err := u.doesPodExceedClusterCapacity(ctx, pod)
+			if err != nil {
+				svc1log.FromContext(ctx).Error("failed to check if pod was unschedulable",
+					svc1log.SafeParam("podName", pod.Name),
```

Comment:
you can capture `podName` and `podNamespace` using `ctx = svc1log.WithLoggerParams(ctx, params...)`